### ISSUE TITLE
Op in first training step is much slower

### ISSUE THREAD

**Issue by jupvfranco** [microsoft/onnxruntime#5715]
**Describe the bug**

When profiling the first training step of a model, one of the operations always takes much longer than the remaining ones. I guess that in some runtimes, it is common to be slower in the beginning of the execution. However, this got my attention because: 

1. This was not the behavior in a commit from a few months ago. The screenshot A below shows the first training step of the MNIST model (cpp implementation) measured with an old commit (https://github.com/jupvfranco/onnxruntime/tree/nvtx -- back then we had our own patch to emit nvtx markers per kernel). The first operator (Gemm) takes longer than the rest of the operations, but this seems to be mainly because of the cuda malloc marked in red. While, this is a 7x slow down w.r.t other operations of the same type, with a most recent version we get more than a 3K slowdown. 
 
2. The operation that is affected is not always the same in different runs. I profiled the first training step of the same model several times, and each time I got a different op. I left a list of operators affected, their duration (in milliseconds, while all the other ops seem to take microseconds), and screenshots below.   

So far, I've only seen this behavior in the first training step. 

Is this an expected behavior? And if so, why does it happen?

**Urgency**
Not urgent

**System information**
OS Platform and Distribution: Ubuntu 18.04.5 LTS
ONNX Runtime installed from: source
ONNX Runtime version: onnxruntime-gpu-1.5.0
GCC/Compiler version: 7.5.0
CUDA/cuDNN version: CUDA version 11.0 (and NCCL 2.7.8)
GPU model and memory: Tesla V100-PCIE-16GB

**To Reproduce**
```
./build.sh --enable_training --use_cuda --cuda_home /usr/local/cuda --cudnn_home /usr/local/cuda --config=Debug --build_wheel  --skip_tests --parallel --enable_nvtx_profile
nvprof --profile-child-processes -o profile%p.prof -- path/to/onnxruntime/build/Linux/Debug/onnxruntime_training_mnist --model_name mnist_gemm --train_data_dir mnist_data --log_dir logs/  --num_train_steps 1 --use_cuda
```
I took the mnist_gemm.onnx file from: https://github.com/microsoft/onnxruntime/files/4553416/mnist.zip

I used NVIDIA's nvvp to check the prof files, but I’ve been able to observe this using the `edge://tracing` tool too. 

**Expected behavior**
I would expect something similar to screenshot A.

**Screenshots**
A) Observed with code from August
![nvtx_branch_behaviour](https://user-images.githubusercontent.com/6861114/98282283-e6574800-1f95-11eb-87ed-acd53ba66a97.png)

Operator affected: ReluGrad (bwd)
Duration: 161 ms
![ReluGrad161](https://user-images.githubusercontent.com/6861114/98282381-10a90580-1f96-11eb-9204-e9448be6b84b.png)

Operator affected: Gemm (bwd) 
Duration: 157 ms
![Gemm157](https://user-images.githubusercontent.com/6861114/98282404-1acb0400-1f96-11eb-85af-39aa4e3c17c7.png)
 
Operator affected: ReduceSumTraining (bwd) 
Duration: 156 ms
![ReduceSumTraining156](https://user-images.githubusercontent.com/6861114/98282410-1ef72180-1f96-11eb-9a86-9a8ae9c16f60.png)

Operator affected: Gemm (bwd) 
Duration: 158 milliseconds
![Gemm158](https://user-images.githubusercontent.com/6861114/98282427-23bbd580-1f96-11eb-9cda-302b45629b5d.png)

**Comment by tlh20** [comment]
Hi Juliana, I discussed this a little with a few folk here.  No-one's had time to dig into it very deeply or repro, but one suspicion is that it's JIT related.  I'm curious about the change in behavior from a prior commit -- is that the exact same toolchain, versions, and h/w?

**Comment by jupvfranco** [comment]
Hi Tim, 

thank you. It makes sense that this is JIT related. 

I believe I used the same VM spec for both versions. The only difference is that back then we didn't find nvtx annotations for individual ops, so we added our own [patch](https://github.com/jupvfranco/onnxruntime/tree/nvtx), but I believe this is unrelated. 

Here's 3 training steps of mnist with a recent version of ort. It is interesting because the first training step is not thousands of times slower than the rest of the steps. However, there is a long cudaEventDestroy in between steps. 
![image](https://user-images.githubusercontent.com/6861114/99280645-7a909d00-2829-11eb-8cb4-64f6d5eee043.png)

Looking at a bert-tiny from August, we also see that the first training step takes longer than the remaining ones. The slowdown is not as bad as in the mnist case though. Could this be related with the size of the model? 
![image](https://user-images.githubusercontent.com/6861114/99278766-6055bf80-2827-11eb-821f-7dea84bce9dc.png)

As we discussed before, I'm not blocked on this. I'll ignore the first training step anyway when profiling. Feel free to close this issue, if you think we shouldn't investigate it further.
